File: Semester-4/FPROG/A11/functionalSum.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

struct Sums {
  Sums() : evenSum(0), oddSum(0), total(0) {}
  Sums(const int &even, const int &odd, const int &total)
      : evenSum(even), oddSum(odd), total(total) {}

  int evenSum;
  int oddSum;
  int total;
};

const bool isEven(const int &value) { return value % 2 == 0; }

const bool isOdd(const int &value) { return value % 2 != 0; }

template <typename T, template <typename...> typename CollectionType>
const T sum(const CollectionType<T> &input, const T &init = 0) {
  return accumulate(input.begin(), input.end(), init);
}

template <typename CollT, typename UnaryPredicate>
const CollT filter(const CollT &input, UnaryPredicate filterFunction) {
  CollT filtered;
  copy_if(input.begin(), input.end(), back_inserter(filtered), filterFunction);
  return filtered;
}

const Sums sums(const vector<int> &numbers) {
  return Sums(sum(filter(numbers, isEven)), sum(filter(numbers, isOdd)),
              sum(numbers));
}

int main() {
  vector<int> numbers = {1, 2, 3, 4, 5, 6, 7, 8};

  Sums theTotals = sums(numbers);
  cout << theTotals.oddSum << ", " << theTotals.evenSum << ", "
       << theTotals.total << endl;
}

File: Semester-4/FPROG/A11/functionalSum.hs
import Data.List (foldl')

-- Define a data structure for sums
data Sums = Sums { evenSum :: Int, oddSum :: Int, totalSum :: Int } deriving (Show)

-- Check if a number is even
isEven :: Int -> Bool
isEven n = n `mod` 2 == 0

-- Check if a number is odd
isOdd :: Int -> Bool
isOdd n = n `mod` 2 /= 0

-- Calculate the sum of a list
sumList :: [Int] -> Int
sumList = foldl' (+) 0

-- Filter a list based on a predicate
filterList :: (a -> Bool) -> [a] -> [a]
filterList = filter

-- Calculate the sums
calculateSums :: [Int] -> Sums
calculateSums numbers = 
    Sums 
        { evenSum = sumList $ filterList isEven numbers
        , oddSum = sumList $ filterList isOdd numbers
        , totalSum = sumList numbers
        }

main :: IO ()
main = do
    -- Define a list of numbers
    let numbers = [1, 2, 3, 4, 5, 6, 7, 8]
    
    -- Calculate the sums
    let result = calculateSums numbers
    
    -- Print the results
    putStrLn $ "Odd Sum: " ++ show (oddSum result) ++ ", Even Sum: " ++ show (evenSum result) ++ ", Total: " ++ show (totalSum result)
